File: DataStructure/DataStructure_CPP/2.Stack and Queue/Queue.cc
#include<iostream>
#include<string>
using namespace std;


class Queue{
private:
    struct Node{
        string item;
        Node *next;
    };
    Node *first, *last;
public:
    Queue(){
        first = NULL;
        last = NULL;
    }
    bool IsEmpty();
    void enqueue(string);
    string dequeue();

};

bool Queue::IsEmpty(){
    return first == NULL;
}

void Queue::enqueue(string i){
    Node *oldlast = last;
    last = new Node();
    last->item = i;
    last->next = NULL;
    if(Queue::IsEmpty()){
        first = last;
    }else{
        oldlast->next = last;
    }


}

string Queue::dequeue(){
    string j = first->item;
    Node *tem = first;
    first = first->next;
    if(IsEmpty()){
        last = NULL;
    }
    delete tem;
    return j;

}

int main(){
    Queue q;
    int i;
    while(1){
        cout<<"1. Enqueue"<<endl;
        cout<<"2. Dequeue"<<endl;
        cin>>i;
        if(i == 1){
            string k;
            cout<<"String: "<<endl;
            cin>>k;
            q.enqueue(k);
        }else if( i == 2 ){
            string s = q.dequeue();
            cout<< s <<endl;
        }
    }

}

File: Algorithms/w2_stack_queue/Queue/Queue.java
package Queue;
public class Queue<T>{
    private class Node<T>{
        T item;
        Node<T> next;
    }
    private Node<T> first, last;
    public boolean isEmpty() {
        return first == null;
    }
    public void enqueue(T i){
        Node<T> oldlast = last;
        last = new Node<T>();
        last.item = i;
        last.next = null;
        if(isEmpty()){
            first = last;
        }else{
            oldlast.next = last;
        }


    }
    public T dequeue(){
        T j = first.item;
        first = first.next;
        if(isEmpty()){
            last = null;
        }
        return j;


    }
    

}
